File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/python_cpp/_378_Convert_Binary_Search_Tree_to_Doubly_Linked_List_Medium.py
'''/**
 * 378
 * Convert Binary Search Tree to Doubly Linked List
 '''


"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None
Definition of Doubly-ListNode
class DoublyListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = self.prev = next
"""

class Solution:
    """
    @param root, the root of tree
    @return: a doubly list node
    """
    def bstToDoublyList(self, root):
        # Write your code here
        dfs = []
        self.getDfs(root, dfs)
        if len(dfs) == 0:
            return None

        head = None
        prev = None
        for val in dfs:
            node = DoublyListNode(val)
            if head is None:
                head = node
            else:
                prev.next = node
            node.prev = prev
            prev = node
        return head

    def getDfs(self, root, dfs):
        if root is None:
            return
        self.getDfs(root.left, dfs)
        dfs.append(root.val)
        self.getDfs(root.right, dfs)

File: JS9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/python_cpp/_378_Convert_Binary_Search_Tree_to_Doubly_Linked_List_Medium.cpp
/**
 * 378
 * Convert Binary Search Tree to Doubly Linked List
 */


/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 * Definition of Doubly-ListNode
 * class DoublyListNode {
 * public:
 *     int val;
 *     DoublyListNode *next, *prev;
 *     DoublyListNode(int val) {
 *         this->val = val;
           this->prev = this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of tree
     * @return: the head of doubly list node
     */
    DoublyListNode* bstToDoublyList(TreeNode* root) {
        // Write your code here
        if (root == NULL)
            return NULL;
        DoublyListNode* node = getDoublyList(root, NULL, NULL);
        while (node->prev != NULL)
            node = node->prev;
        return node;
    }
    DoublyListNode* getDoublyList(TreeNode* root,
                                  DoublyListNode* prev,
                                  DoublyListNode* next) {
        if (root == NULL)
            return NULL;
        DoublyListNode* node = new DoublyListNode(root->val);
        if (root->left == NULL) {
            node->prev = prev;
            if (prev != NULL)
                prev->next = node;
        } else
            getDoublyList(root->left, prev, node);
        if (root->right == NULL) {
            node->next = next;
            if (next != NULL)
                next->prev = node;
        } else
            getDoublyList(root->right, node, next);

        return node;
    }
};


